### PR TITLE
Feature/rp 130 alt score ui

### PR DESCRIPTION
**Pull request by esotrope**
Introduces actual number crunching for alt scores

image

### CODE REVIEW

File: webapp/src/main/webapp/src/app/assessments/assessment-exam.mapper.ts [hunk 1]
```diff
@@ -7,12 +7,20 @@ import { ExamItemScore } from './model/exam-item-score.model';
...
 import { byNumber } from '@kourge/ordering/comparator';
-import { ClaimScore } from './model/claim-score.model';
 import { Student } from '../student/model/student.model';
 import { Utils } from '../shared/support/support';
 import { DefaultSchool, School } from '../shared/organization/organization';
 import { TargetScoreExam } from './model/target-score-exam.model';
 import { Target } from './model/target.model';
+import { ScaleScore } from '../exam/model/scale-score';
+
+function toScaleScore(serverScaleScore: any): ScaleScore {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/assessment-exam.mapper.ts]: removed concept of "ClaimScore" in favor of "ScaleScore" which was the same thing

File: webapp/src/main/webapp/src/app/assessments/assessment-exam.mapper.ts [hunk 2]
```diff
@@ -105,8 +113,11 @@ export class AssessmentExamMapper {
     exam.schoolYear = serverExam.schoolYear;
     exam.transfer = serverExam.transfer;
     exam.school = this.createSchool(serverExam.school);
+    exam.alternateScaleScores = (serverExam.altScaleScores || []).map(
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/assessment-exam.mapper.ts]: added mapping for alt scores on exams

File: webapp/src/main/webapp/src/app/assessments/model/claim-score.model.ts [hunk 3]
```diff
@@ -1,14 +1,7 @@
 import { ExamStatisticsLevel } from './exam-statistics.model';
 
-export class ClaimScore {
-  level: number;
-  standardError: number;
-  score: number;
-}
-
 export class ClaimStatistics {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/model/claim-score.model.ts]: i would like to deprecate this and `ExamStatistics` and replace it with what i have for alt scores in a followup refactor

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts [hunk 4]
```diff
@@ -521,6 +527,15 @@ export class AssessmentResultsComponent implements OnInit {
     this.exams = this.filterExams();
     this.statistics =
       this.subjectDefinition != null ? this.calculateStats() : null;
+
+    // compute table when ready
+    if (this.subjectDefinition != null && this.exams != null) {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: this will be recomputed everytime the input exams change (so when advanced filters change etc)

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts [hunk 5]
```diff
@@ -540,29 +555,31 @@ export class AssessmentResultsComponent implements OnInit {
...
   }
 
+  // todo use scoreStatistics
   private calculateStats(): ExamStatistics {
     const stats = new ExamStatistics();
 
-    let scores = this.examCalculator
-      .getOnlyScoredExams(this.exams)
+    const scores = this.exams
+      .filter(exam => exam && exam.score) // should be score != null but original examCalculator has it this way
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: @malaffoon currently we don't count exams with `score === 0` in the overall chart but we do count it in the "result count"
so if we had `Exam{score=0}` we would show:
`1 result` and `0%` for all performance levels.

this doesn't feel right to me. thoughts?

File: webapp/src/main/webapp/src/app/assessments/results/exam-statistics-calculator.spec.ts [hunk 6]
```diff
@@ -18,70 +17,23 @@ describe('Exam Calculator', () => {
     expect(actual.length).toBe(2);
   });
 
-  it('should calculate the average', () => {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/exam-statistics-calculator.spec.ts]: moved tests along with methods to `score-statistics.ts` and `*.spec.ts`

File: webapp/src/main/webapp/src/app/assessments/results/scale-score.service.spec.ts [hunk 7]
```diff
@@ -1,143 +0,0 @@
-import { inject, TestBed } from '@angular/core/testing';
-import { ScaleScoreService } from './scale-score.service';
-import { ExamStatisticsLevel } from '../model/exam-statistics.model';
-
-describe('ScaleScoreService', () => {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/scale-score.service.spec.ts]: this service was not being used anywhere and seemed to be an attempt at what we now call the method `roundPercentages`

File: webapp/src/main/webapp/src/app/assessments/results/target-statistics-calculator.ts [hunk 8]
```diff
@@ -181,28 +240,42 @@ export class TargetStatisticsCalculator {
...
+      return TargetReportingLevel.NoResults;
+    }
+
+    const scored = scores.filter(value => value != null);
 
     return this.mapTargetScoreDeltaToReportingLevel(
-      this.examStatisticsCalculator.calculateAverage(scores),
-      this.examStatisticsCalculator.calculateStandardErrorOfTheMean(scores)
+      average(scored),
+      standardErrorOfMean(scores)
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/target-statistics-calculator.ts]: these are pure functions and do not need the overhead of a service wrapper that needs to be injected and instantiated in order to test the method

File: webapp/src/main/webapp/src/app/dashboard/group-dashboard/group-assessment-card.component.spec.ts [hunk 9]
```diff
@@ -1,29 +1,25 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/dashboard/group-dashboard/group-assessment-card.component.spec.ts]: formatting only

File: webapp/src/main/webapp/src/app/dashboard/group-dashboard/group-assessment-card.component.ts [hunk 10]
```diff
@@ -49,8 +47,10 @@ export class GroupAssessmentCardComponent implements OnInit {
   }
 
   ngOnInit() {
-    this.percents = this.measuredAssessment.studentCountByPerformanceLevel.map(x => Math.round(x.percent));
-    this.dataWidths = this.examCalculator.getDataWidths(
+    this.percents = this.measuredAssessment.studentCountByPerformanceLevel.map(
+      x => Math.round(x.percent)
+    );
+    this.dataWidths = roundPercentages(
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/dashboard/group-dashboard/group-assessment-card.component.ts]: only change was to cut over to this method

File: webapp/src/main/webapp/src/app/exam/component/score-table/score-tables.ts [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+  const metadata = ScoreTypeMetadataByType.get(scoreType);
+
+  const examScaleScores = exams
+    .map(exam => metadata.scaleScores(exam))
+    .filter(
+      scaleScores =>
+        !isNullOrEmpty(scaleScores) &&
+        scaleScores.every(
+          scaleScore =>
+            scaleScore != null &&
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/exam/component/score-table/score-tables.ts]: open item on whether or not we ignore score === 0 here too

File: webapp/src/main/webapp/src/app/exam/model/score-statistics.ts [hunk 12]
```diff
@@ -62,8 +47,174 @@ export interface PerformanceLevelScore {
...
+  return rounded.map(({ roundedPercent }) => roundedPercent);
+}
+
+/**
+ * Calculates the performance level scores given the exam scale score levels and subject/assessment score levels
+ *
+ * @param examLevels The exams' scale score levels
+ * @param scoreLevels The subject/assessment's performance levels
+ */
+export function performanceLevelScores(
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/exam/model/score-statistics.ts]: this and the below method are net-new, the above ones are established and i just renamed them and brought their tests over to this module

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts (hunk 5 above)
Comment (malaffoon) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: Bit of a corner case. I can see arguments for both sides. If the result count is meant to represent the tests taken then it should be `1`. If it is meant to represent the tests contributing to the summary then, yeah, it should be `0`. Hmm, we might have to ask.

Comment (esotrope) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: is an unscored result even worth seeing in the UI? feels like there should be a filter that defaults to only show scored results and you have to go fiddling to find the unscored. not sure how its being used today though. it might help to find out
